### DISCUSSION TITLE
Replacing substructures at 2+ bonding locations

### DISCUSSION THREAD

**Discussion by jmohin** [rdkit/rdkit#4488] [General]
When programmatically building molecules, it's often desirable to replace a fused ring group.  For instance, if I want to replace the benzene ring in indole, I can easily identify it's substructure using GetSubstructureMatch() but when I replace it, it's singly-bonded to the remaining pyrrole group.  What functionality, if any, allows me to specify HOW I substitute a substructure?

**Comment by ptosco** [comment]
@jmohin Could you please share the code you are currently using? Unless the ring you wish to replace benzene with is symmetric (e.g., cyclohexane), you will also need to indicate which bond of the incoming ring you would like to condense with, or alternatively enumerate all possible condensation products. For example, replacing benzene with pyridine could be done in 6 different ways, leading to 6 different condensation products.

**Comment by jmohin** [comment]
Here's an example which illustrates the problem:
`indole=Chem.MolFromSmiles('C12=C(C=CN2)C=CC=C1')
Chem.ReplaceSubstructs(indole,Chem.MolFromSmiles('c1ccccc1'),Chem.MolFromSmiles('c1ccsc1'))[0]`

Yields an invalid/unkekulized molecule wherein the thiophene is singly-bonded to a fragment from the indole.  

I know there's six ways to attach a pyridine to the pyrrole resulting from deleting benzene in indole -- lets assume I know the atom indices I want to bond to on both the pyrrole and incoming pyridine.  What functionality would I use?

**Comment by jmohin** [comment]
I'm aware of the option to ReplaceSubstructs:
'replacementConnectionPoint: (optional) index of the atom in the replacement that the bond should be made to.'
but it only accepts a single integer -- I would like it to accept a list of tuples where each tuple is a pair of indices corresponding to the original molecule and index in the replacement molecule I want to bond at.

**Reply by jmohin** [reply]
I found a way to do this using reaction SMARTs, for instance, making naphthalene in to benzothiophene:
```
napth=Chem.MolFromSmiles('c1c2ccccc2ccc1')
rxn = AllChem.ReactionFromSmarts(
    f'[*:1]c1ccccc1[*:2]>>[*:1]c1cscc1[*:2]')
reacts=rxn.RunReactants((napth,))

```
however,  two questions: 
1.) I get 4 molecules when I run this reaction; I'm assuming those are the combinatorial attachment possibilies, but all 4 molecules are identical.  Can RDkit determine identical molecules?
2.) if I wanted to react twice to get bithiophene, how would I specify the number of times I want to match/replace?

**Comment by fdaqin** [comment]
@jmohin Does the Chem.ReplaceCore() help?
